tests/test_covar2d.py

```diff
@@ -154,3 +156,59 @@ def testGetCWFCoeffsClean(self):
         self.assertTrue(
             np.allclose(results, self.coeff_cwf_clean, atol=utest_tolerance(self.dtype))
         )
+
+    def testGetCWFCoeffsCleanCTF(self):
+        """
+        Test case of clean images (coeff_clean and noise_var=0)
+        while using a non Identity CTF.
+
+        This case may come up when a developer switches between
+        clean and dirty images.
+        """
+
+        coeff_cwf = self.cov2d.get_cwf_coeffs(
+            self.coeff_clean, self.h_ctf_fb, self.h_idx, noise_var=0
+        )
+
+        # Reconstruct images from output of get_cwf_coeffs
+        img_est = self.basis.evaluate(coeff_cwf)
+        # Compare with clean images
+        delta = np.mean(np.square((self.imgs_clean - img_est).asnumpy()))
+        self.assertTrue(delta < 0.02)
+
+    def testGetCWFCoeffsCTFargs(self):
+        """
+        Test we raise when user supplies incorrect CTF arguments
+        """
+
+        with raises(RuntimeError):
+            _ = self.cov2d.get_cwf_coeffs(
+                self.coeff, self.h_ctf_fb, None, noise_var=self.noise_var
+            )
+
+    @parameterized.expand(
+        [(None,), "frobenius_norm", "operator_norm", "soft_threshold", "None"]
+    )
+    def testShrinkers(self, shrinker):
+        """Test all the shrinkers we know about run, and we raise otherwise."""
+
+        if shrinker == "None":
+            with raises(AssertionError):
+                _ = self.cov2d.get_covar(
+                    self.coeff_clean, covar_est_opt={"shrinker": shrinker}
+                )
+            return
+
+        results = np.load(
+            os.path.join(DATA_DIR, "clean70SRibosome_cov2d_covar.npy"),
+            allow_pickle=True,
+        )
+
+        covar_coeff = self.cov2d.get_covar(
+            self.coeff_clean, covar_est_opt={"shrinker": shrinker}
+        )
+
+        for im, mat in enumerate(results.tolist()):
+            self.assertTrue(
+                np.allclose(mat, covar_coeff[im], atol=utest_tolerance(self.dtype))
```

Comment:
So since `noise_var = 0`, all these shrinkers should be no-ops, is that what we're testing?

Reply:
I'm testing they should not crash.

Reply:
Got it.